File: src/main/java/no/nav/pensjon/innsyn/main/WorksheetProducer.java
package no.nav.pensjon.innsyn.main;

import no.nav.pensjon.innsyn.domain.desc.*;
import no.nav.pensjon.innsyn.domain.map.*;
import no.nav.pensjon.innsyn.domain.source.*;
import no.nav.pensjon.innsyn.entity.EntitySupport;
import no.nav.pensjon.innsyn.poi.ss.DataTransferrer;
import no.nav.pensjon.innsyn.poi.xssf.ExcelWorkbookCreator;
import no.nav.pensjon.innsyn.sink.OutputStreamCreator;
import no.nav.pensjon.innsyn.sql.DbConnection;

import java.io.IOException;
import java.sql.Connection;
import java.sql.SQLException;
import java.util.List;

public class WorksheetProducer {

    public static void produceWorksheet(String fnr, OutputStreamCreator outputStreamCreator) throws IOException {
        try (Connection connection = DbConnection.get()) {
            List<EntitySupport<?>> entitySupports = List.of(
                    new EntitySupport<>(
                            new DbBeholdningGetter(fnr, connection),
                            new BeholdningDescriptor(),
                            BeholdningRowFiller::setCellValues),
                    new EntitySupport<>(
                            new DbInntektGetter(fnr, connection),
                            new InntektDescriptor(),
                            InntektRowFiller::setCellValues),
                    new EntitySupport<>(
                            new DbDagpengerGetter(fnr, connection),
                            new DagpengerDescriptor(),
                            DagpengerRowFiller::setCellValues),
                    new EntitySupport<>(
                            new DbFppAfpGetter(fnr, connection),
                            new FppAfpDescriptor(),
                            FppAfpRowFiller::setCellValues),
                    new EntitySupport<>(
                            new DbOmsorgGetter(fnr, connection),
                            new OmsorgDescriptor(),
                            OmsorgRowFiller::setCellValues),
                    new EntitySupport<>(
                            new DbForstegangstjenesteGetter(fnr, connection),
                            new ForstegangstjenesteDescriptor(),
                            ForstegangstjenesteRowFiller::setCellValues),
                    new EntitySupport<>(
                            new DbOpptjeningGetter(fnr, connection),
                            new OpptjeningDescriptor(),
                            OpptjeningRowFiller::setCellValues));

            DataTransferrer.transferEntitiesToWorkbook(
                    entitySupports,
                    new ExcelWorkbookCreator(),
                    outputStreamCreator);
        } catch (SQLException e) {
            throw new IOException("Failed to get DB connection: " + e.getMessage(), e);
        }
    }
}

File: src/main/kotlin/no/nav/pensjon/innsyn/main/WorksheetProducer.kt
package no.nav.pensjon.innsyn.main

import no.nav.pensjon.innsyn.domain.*
import no.nav.pensjon.innsyn.domain.desc.*
import no.nav.pensjon.innsyn.domain.map.DomainRowFiller
import no.nav.pensjon.innsyn.domain.source.*
import no.nav.pensjon.innsyn.entity.EntitySupport
import no.nav.pensjon.innsyn.poi.ss.SheetFiller
import no.nav.pensjon.innsyn.sql.DbConnection
import org.apache.poi.xssf.usermodel.XSSFWorkbook
import java.io.IOException
import java.sql.SQLException

object WorksheetProducer {
    fun producePOPPWorksheet(fnr: String) = try {
        DbConnection.get().use { connection ->
            XSSFWorkbook().use {
                it.populate(
                        EntitySupport(
                                DbBeholdningGetter(fnr, connection),
                                BeholdningDescriptor,
                                DomainRowFiller(Beholdning::class)::setCellValues),
                        EntitySupport(
                                DbInntektGetter(fnr, connection),
                                InntektDescriptor,
                                DomainRowFiller(Inntekt::class)::setCellValues),
                        EntitySupport(
                                DbDagpengerGetter(fnr, connection),
                                DagpengerDescriptor,
                                DomainRowFiller(Dagpenger::class)::setCellValues),
                        EntitySupport(
                                DbFppAfpGetter(fnr, connection),
                                FppAfpDescriptor,
                                DomainRowFiller(FppAfp::class)::setCellValues),
                        EntitySupport(
                                DbOmsorgGetter(fnr, connection),
                                OmsorgDescriptor,
                                DomainRowFiller(Omsorg::class)::setCellValues),
                        EntitySupport(
                                DbForstegangstjenesteGetter(fnr, connection),
                                ForstegangstjenesteDescriptor,
                                DomainRowFiller(Forstegangstjeneste::class)::setCellValues),
                        EntitySupport(
                                DbOpptjeningGetter(fnr, connection),
                                OpptjeningDescriptor,
                                DomainRowFiller(Opptjening::class)::setCellValues)
                )
                it
            }
        }
    } catch (e: SQLException) {
        throw IOException("Failed to get DB connection: ${e.message}", e)
    }

    private fun XSSFWorkbook.populate(vararg entitySupports: EntitySupport<*>) =
            entitySupports.map { SheetFiller(it) }
                    .forEach { it.populateSheet(this) }
}